### ISSUE TITLE
Improve message for `Mismatching_for_pack` errors

### ISSUE THREAD

**Issue by jchavarri** [ocaml/ocaml#12581]
I stumbled upon this kind of error:

```
ocamlfind ocamlopt -c -for-pack Camlp4.Struct -g -safe-string -w Z -I camlp4/import -package camlp-streams -pp 'camlp4/boot/camlp4boot.byte -D OPT' -I camlp4/Camlp4/Struct -I camlp4/Camlp4 -I camlp4/config -I camlp4 -o camlp4/Camlp4/Struct/Loc.cmx camlp4/Camlp4/Struct/Loc.ml
File "camlp4/Camlp4/Struct/Loc.ml", line 1:
Error: camlp4/Camlp4/ErrorHandler.cmx
       was built with -for-pack Camlp4, but the
       current unit Loc is built with -for-pack Camlp4.Struct
Command exited with code 2.
```

It looks like the compiler already knows:
- the module where the problem is happening
- the referenced module that is failing
- the expected pack
- the current pack

Maybe the error message could include some recommended steps? For example, in the message above, it would be something like:

```
Error: camlp4/Camlp4/ErrorHandler.cmx
       was built with -for-pack Camlp4, but the
       current unit Loc is built with -for-pack Camlp4.Struct
To fix the problem, try namespacing the appearances of ErrorHandler inside Loc, e.g. Camlp4.ErrorHandler
```

**Comment by gasche** [comment]
To my knowledge, for-pack is extremely rarely used these days (only in ocamlbuild projects?). Do we want to spend maintainer time on improving its ergonomics?

I found your message a bit perplexing -- I am not sure what the error means, your explanation is obscure as well (you have a build error in Loc.ml that you propose to fix by changing code in ErrorHandler.ml?), and you seem to be referring to your own experimental changes to Camlp4 (maybe?), without much context to reproduce the issue or look at the faulty code.

The compiler, on the other hand, has a code comment about this error which I find clearer than both the current error message or your suggestion:

```
            (* Linking to a compilation unit expected to go into a
               pack (ui_for_pack = Some ...) is possible only from
               inside the same pack, but it is perfectly ok to link to
               an unit outside of the pack. *)
```

Maybe one would just need to recall this inside the error, or at least point to the manual chapter where this is mentioned explicitly, if any?

**Comment by lthls** [comment]
I'll need to experiment a bit, but I think this looks like we're not considering `Camlp4.Struct.Loc` as included in `Camlp4`. @lukemaurer recently rewrote a good chunk of the compilation unit infrastructure on the Jane Street branch of the compiler, and I suspect that with his work references to `ErrorHandler` from `Loc` would be allowed as the pack prefixes are compatible.

**Comment by jchavarri** [comment]
@gasche Yes, to add extra context, I ran into this error while trying to upgrade camlp4 [to support 5.1](https://github.com/camlp4/camlp4/pull/166).

In that branch, the bytecode build compiles fine, but the native build fails with the error I pasted on the issue description.

> To my knowledge, for-pack is extremely rarely used these days (only in ocamlbuild projects?). Do we want to spend maintainer time on improving its ergonomics?

I can see how it is uncommon to run into this issue these days. If that is the case, feel free to close the issue, I don't think anyone wants is to spend maintainers time on improvements that will barely have any impact.

**Comment by jchavarri** [comment]
> you have a build error in Loc.ml that you propose to fix by changing code in ErrorHandler.ml?

Apologies, there was a mistake in the original description. Should be fixed now.

**Comment by gasche** [comment]
My understanding of what is going wrong:

- in 5.1 we included a breaking change, an old PR from @chambart, that was only supposed to "uncover hidden build system bugs": #1391
- that change broke the build of Camlp4 (which uses namespaced modules the old way, with ocamlbuild and -for-pack)
- @jchavarri is now scrambling to fix the build of Camlp4 by performing ugly hacks in the codebase ( https://github.com/camlp4/camlp4/pull/166 )

We have not analyzed the breakage properly and we don't know if the use-case of Camlp4 was in fact reasonable. I think we should seriously consider reverting the change, instead of improving error messages.

(cc @chambart @lthls @Octachron)

**Comment by gasche** [comment]
If I understand correctly (I haven't touched the Camlp4 codebase in a good while, apologies), Camlp4 uses two nested packs:

- Camlp4, with some modules
- Camlp4.Struct, with some modules

The problem is that some of the Camlp4.Struct modules depend on some of the Camlp4 modules that are "earlier" than Struct in dependency order.

It is not obvious to me that this is a morally wrong thing to do, and it has been supported for decades. Could the proponent of the change in #1391 explain why we should break this now? (For example, does this particular case of "mismatched for-pack" create issues with flambda?

If we consider that this is a wrong thing to do, then what fix strategy are we recommending to users? For now the fix that @jchavarri is performing is, whenever a module from `Struct` depends on a module `Camlp4.Foo`, to move the module to `Camlp4.Struct.Foo`. This does not sound like a particularly palatable solution to me, it breaks the organization principles that were in place before.

(One strategy we could consider is to stop using packs and perform everything manually: rename your files into `Camlp4__Foo`,  `Camlp4__Struct__Bar`, with a flat hierarchy of encoded names, and then rebuild the `Camlp4` and `Camlp4.Struct` modules using module aliases. At least the documentation-generation tools, etc. would see a now-standard wrapped layout instead of extra hackish modules in Camlp4.Struct. I don't know if this solution is workable in practice within the context of an ocamlbuild build system.)

**Comment by lthls** [comment]
I suspect that we did not properly consider the case of nested packs. I will send a patch that adds back support for this case, as `camlp4` might not be the only impacted package.

**Comment by lthls** [comment]
This should have been fixed by #12609.
